gitbootcamp.rst

```diff
@@ -0,0 +1,268 @@
+.. _gitbootcamp:
+
+Git Bootcamp and Cheat Sheet
+============================
+
+In this section, we'll go over some commonly used git commands that are
+relevant to CPython's workflow.
+
+.. contents::
+
+
+Forking CPython GitHub Repository
+---------------------------------
+
+You'll only need to do this once.
+
+1. Go to https://github.com/python/cpython.
+
+2. Press ``Fork``.
+
+3. When asked where to fork the repository, choose to fork it to your username.
+
+4. A fork will be created at https://github.com/<username>/cpython.
+
+
+Cloning The Forked CPython Repository
+-------------------------------------
+
+From your command line::
+
+   $ git clone git@github.com:<username>/cpython.git
+   $ cd cpython
+   $ git remote add upstream git@github.com:python/cpython.git
+
+
+Listing the Remote Repositories
+-------------------------------
+
+::
+
+   $ git remote -v
+
+
+Creating and Switching Branches
+-------------------------------
+
+.. note::
+   Never commit directly to the ``master`` branch.
+
+Create a new branch::
+
+   $ git checkout -b some-branch master # creates a new branch off master
+
+which is equal to::
+
+   $ git branch some-branch master # create 'some-branch' off 'master', without checking it out
+   $ git checkout some-branch # check out 'some-branch'
+
+To find out which branch you are in now::
+
+   $ git branch
+
+The current branch will have an asterisk next to the branch name.  Note, this
+will list all of your local branches.
+
+To list all the branches, including the remote branches::
+
+   $ git branch -a
+
+To switch to a different branch::
+
+   $ git checkout another-branch-name
+
+
+Delete Local Branch
+-------------------
+
+To delete branch that you no longer need::
+
+   $ git branch -D branch-to-delete
+
+
+Staging and Committing Files
+----------------------------
+
+1. To show the current changes::
+
+      $ git status
+
+2. To stage the files to be included in your commit::
+
+      $ git add /path/to/file1 path/to/file2 path/to/file3
+
+3. To commit the files that have been staged (done in step 2)::
+
+      $ git commit -m "This is the commit message. Prefix it with bpo-XXXX."
+
+
+Reverting Changes
+-----------------
+
+To revert changes to a file that has not been committed yet::
+
+    $ git checkout path/to/file
+
+If the change has been committed, and now you want to reset it to whatever
+the origin is at::
+
+   $ git reset --hard HEAD
```

Comment:
There are two more scenarios where I used `hg revert` that are not covered here:
1. how to revert all changes in the working copy (i.e. the equivalent of `hg revert -a` -- `git reset`? `git reset --hard`?).
2. how to revert the fix while leaving the test to verify if they fail without the fix.

The second case works differently since we are using PRs.  With HG I could apply a patch and do `hg revert Lib/` to revert all the changes in `Lib` while leaving the tests unchanged.  I think the git equivalent would be something like `git checkout master Lib/`, after I pulled a PR as described in a later section.